internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,312 @@
+	// key included must verify the signature). We do this here as a
+	// sanity check since we are about to pair this certificate and use
+	// it persisted certificate in exactly this way for future client
+	// credential checks. Note that the TLS handshake already proved
+	// that the client has access to the private key by verifying the
+	// handshake transcript signature using the public key in this
+	// certificate.
+	roots := x509.NewCertPool()
+	roots.AddCert(clientCert)
+	opts := x509.VerifyOptions{
+		Roots: roots,
+		// We only support verifying client TLS certificates.
+		KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+	_, err := clientCert.Verify(opts)
+	if err != nil {
+		return fmt.Errorf("cannot verify client certificate signature: %w", err)
+	}
+
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	existingIdentities := m.state.Identities()
+
+	for _, identity := range existingIdentities {
+		if identity.Cert == nil || identity.Cert.X509 == nil {
+			// Not valid certificate identity.
+			continue
+		}
+
+		if identity.Cert.X509.Equal(clientCert) {
+			// This identity is already added so in this special
+			// case we complete the pairing request without adding
+			// it again with a new username.
+			m.paired = true
+			m.state.Set(pairedStateKey, m.paired)
+
+			return nil
+		}
+	}
+
+	username, err := generateUniqueUsername(existingIdentities)
+	if err != nil {
+		return fmt.Errorf("cannot create new identity username: %w", err)
+	}
+
+	newIdentity := &state.Identity{
+		Access: state.AdminAccess,
+		Cert:   &state.CertIdentity{X509: clientCert},
+	}
+
+	err = m.state.AddIdentities(map[string]*state.Identity{
+		username: newIdentity,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot add identity: %w", err)
+	}
+
+	m.paired = true
+	m.state.Set(pairedStateKey, m.paired)
+
+	return nil
+}
+
+// PairingWindowEnabled returns whether the pairing window is currently enabled.
+func (m *PairingManager) PairingWindowEnabled() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.enabled
+}
+
+// stopTimer stops the current timer if it exists
+func (m *PairingManager) stopTimer() {
+	if m.timer != nil {
+		m.timer.Stop()
+	}
+}
+
+// startTimer starts a new timer that will disable the pairing window after
+// the given timeout. If the timer exists, we will reuse the timer by
+// resetting it with a new duration.
+func (m *PairingManager) startTimer(timeout time.Duration) {
+	// If timer already exists, just reset it with the new timeout
+	if m.timer != nil {
+		m.timer.Reset(timeout)
+		return
+	}
+
+	m.timer = timeAfterFunc(timeout, func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		m.enabled = false
+	})
+}
+
+// EnablePairing requests the pairing manager to enable the pairing window.
+func (m *PairingManager) EnablePairing(timeout time.Duration) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// If the pairing window is already enabled, reset the timeout duration
+	if m.enabled {
+		m.startTimer(timeout)
```

Comment:
There's a race condition here:

1) Test m.enabled, and it's true
2) Previously scheduled timer is called, blocks on mu.Lock 
3) startTimer resets the timer, schedules new function call
4) This function returns, unblocks mu.Lock
5) Previously scheduled timer acquires mu.Lock, sets enabled to false

Reply:
@niemeyer Latest implementation tries using the ensure feature, as we discussed. What do you think ?